std/traits.d

```diff
@@ -8590,3 +8591,83 @@ enum isCopyable(S) = is(typeof(
     static assert(isCopyable!int);
     static assert(isCopyable!(int[]));
 }
+
+/++
+ + Determines if `F` is a strongly `pure` function. This is true when :
+ + $(UL
+ +   $(LI Each parameter is `immutable`, or can be converted automatically to `immutable`
+ +   (i.e. has no mutable indirections) and is passed by value)
+ +   $(LI The Return Type is `immutable` or can be impicitly converted to `immutable`)
+ + )
+ + Params:
+ + F = The function to be tested
+ +
+ + Returns:
+ + `true` if the function is strongly `pure`, `false` if otherwise.
+ + ++/
+template isStronglyPure(F...)
+if (F.length == 1 && isCallable!F)
+{
+    bool isStronglyPure()
+    {
+        alias STC = ParameterStorageClass;
+
+        if (!hasFunctionAttributes!(F, "pure"))
+            return false;
+
+        static foreach (i, param; Parameters!F)
+        {
+            static if (is(param == immutable))
+            {}
+            else static if (isRefType!param || hasAliasing!param
+                    || ParameterStorageClassTuple!F[i] == STC.ref_
+                    || ParameterStorageClassTuple!F[i] == STC.out_)
+                return false;
+        }
```

Comment:
Looks like we're missing a nice `hasParametersAttribute!(param, "ref", "out")` (a-la `hasFunctionAttributes`) API :/